drivers/crypto/crypto_it8xxx2_sha.c

```diff
@@ -0,0 +1,221 @@
+	uint32_t key;
+
+	__unused uint8_t hash_ctrl;
+
+	/*
+	 * Since W field on it8xxx2 requires big-endian format, change byte
+	 * order before computing hash.
+	 */
+	for (int i = 0; i < SHA_SHA256_BLOCK_LEN_WORDS; i++) {
+		chip_ctx.w_sha[i] = sys_cpu_to_be32(chip_ctx.w_sha[i]);
+	}
+	/*
+	 * Global interrupt is disabled because the CPU cannot access memory
+	 * via the DLM (Data Local Memory) bus while HW module is computing
+	 * hash.
+	 */
+	key = irq_lock();
+	hash_ctrl = sys_read8((uintptr_t)&sha_regs->SHA_HASHCTRLR);
+	sys_write8(hash_ctrl | IT8XXX2_SHA_START_SHA256,
+			(uintptr_t)&sha_regs->SHA_HASHCTRLR);
+	hash_ctrl = sys_read8((uintptr_t)&sha_regs->SHA_HASHCTRLR);
+	irq_unlock(key);
+
+	chip_ctx.w_input_index = 0;
+}
+
+static int it8xxx2_hash_handler(struct hash_ctx *ctx, struct hash_pkt *pkt,
+				bool finish)
+{
+	uint32_t rem_len = pkt->in_len;
+	uint32_t in_buf_idx = 0;
+
+	while (rem_len--) {
+		chip_ctx.w_input[chip_ctx.w_input_index++] =
+						pkt->in_buf[in_buf_idx++];
+		if (chip_ctx.w_input_index >= SHA_SHA256_BLOCK_LEN) {
+			it8xxx2_sha256_module_calculation();
+		}
+	}
+	chip_ctx.total_len += pkt->in_len;
+
+	if (finish) {
+		uint32_t *ob_ptr = (uint32_t *)pkt->out_buf;
+
+		/* Pre-processing (Padding) */
+		memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		chip_ctx.w_input[chip_ctx.w_input_index] = 0x80;
+
+		if (chip_ctx.w_input_index >= 56) {
+			it8xxx2_sha256_module_calculation();
+			memset(&chip_ctx.w_input[chip_ctx.w_input_index],
+			0, SHA_SHA256_BLOCK_LEN - chip_ctx.w_input_index);
+		}
+		chip_ctx.w_sha[15] = sys_cpu_to_be32(chip_ctx.total_len * 8);
+		it8xxx2_sha256_module_calculation();
+
+		for (int i = 0; i < SHA_SHA256_HASH_LEN_WORDS; i++) {
+			ob_ptr[i] = sys_be32_to_cpu(chip_ctx.h[i]);
+		}
+
+		it8xxx2_sha256_init(false);
+	}
+
+	return 0;
+}
+
+static int it8xxx2_hash_session_free(const struct device *dev,
+				struct hash_ctx *ctx)
+{
+	it8xxx2_sha256_init(false);
+
+	return 0;
+}
+
+static inline int it8xxx2_query_hw_caps(const struct device *dev)
+{
+	return (CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS);
+}
+
+static int it8xxx2_hash_begin_session(const struct device *dev,
+				struct hash_ctx *ctx, enum hash_algo algo)
+{
+	if (algo != CRYPTO_HASH_ALGO_SHA256) {
+		LOG_ERR("Unsupported algo");
+		return -EINVAL;
+	}
+
+	if (ctx->flags & ~(it8xxx2_query_hw_caps(dev))) {
+		LOG_ERR("Unsupported flag");
+		return -EINVAL;
+	}
+
+	it8xxx2_sha256_init(false);
+	ctx->hash_hndlr = it8xxx2_hash_handler;
+
+	return 0;
+}
+
+static int it8xxx2_sha_init(const struct device *dev)
+{
+	struct it8xxx2_sha_regs *const sha_regs = IT8XXX2_SHA_REGS_BASE;
+
+	it8xxx2_sha256_init(true);
+	/* Configure base address register for W and H */
+	sha_regs->SHA_HBADDR = ((uint32_t)&chip_ctx >> 6) & 0xfc;
+	/* Configure base address register for K */
+	sha_regs->SHA_HBADDR2 = ((uint32_t)&chip_ctx.k >> 6) & 0xfc;
```

Comment:
No, these two registers are required for calculations.
We only need to configure them once during initialization.

Reply:
right, but if these are registers why not using `sys_write32` here as well?

Reply:
Since drivers for it8xxx2 use register structure to access registers, I implemented this driver in the same style.
And the compiler will keep these lines because these registers are declared with volatile keyword.
What is the advantage of using sys_write8 here? Additional barriers will be added?

Reply:
correct. We tend to write a driver abstracted from the architecture so using IO helpers helps when different architectures have different memory models / barrier requirements.

Reply:
ack.